lib/Dialect/FIRRTL/Transforms/Dedup.cpp

```diff
@@ -0,0 +1,1009 @@
+//===- Dedup.cpp - FIRRTL module deduping -----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements FIRRTL module deduplication.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/SHA256.h"
+
+using namespace circt;
+using namespace firrtl;
+using hw::InnerRefAttr;
+
+/// This stores a mapping from a module name to every NLA that it particiapates
+/// in.
+using NLAMap = DenseMap<Attribute, std::vector<NonLocalAnchor>>;
+
+NLAMap createNLAMap(CircuitOp circuit) {
+  DenseMap<Attribute, std::vector<NonLocalAnchor>> nlaMap;
+  for (auto nla : circuit.getBody()->getOps<NonLocalAnchor>()) {
+    for (auto element : nla.namepath()) {
+      if (auto moduleRef = element.dyn_cast<FlatSymbolRefAttr>())
+        nlaMap[moduleRef.getAttr()].push_back(nla);
+      else
+        nlaMap[element.cast<InnerRefAttr>().getModule()].push_back(nla);
+    }
+  }
+  return nlaMap;
```

Comment:
Isn't this precisely the same information stored on the modules in the class=circt.nonlocal?

Reply:
Since the `circt.nonlocal` is stored on instances (not the module) its more equivalent to:
```c++
SmallVector getParticipatingNLAs(Module mod) {
  for (instance : instanceGraph[mod].uses) {
    for (anno : AnnotationSet(instance->getInstance()) {
      if (anno.class == "circt.nonlocal") {
        results.push_back(symbolTable[anno.getMember("circt.nonlocal"))
      }
    }
    return results;
  }
```